api-docs: Update description for update_message_flags add event.

laurynmm opened the pull request:
Updates the description of the `update_message_flags op: add` event in the get events documentation; see [relevant CZO discussion](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/unreads.3A.20messages.20from.20muted.20users.3F/near/1660912).

@gnprice and @chrisbobbe mentioning you here for a review these changes. I did make a few small edits and added some more links to the text drafted in the conversation in CZO (linked above).

**Screenshots and screen captures:**

<details>
<summary>Updated update_message_flags op: add description</summary>

[Current documentation](https://zulip.com/api/get-events#update_message_flags-add)
![Screenshot from 2023-10-24 15-21-00](https://github.com/zulip/zulip/assets/63245456/257ac664-50e4-40ac-a206-57a392a759b4)

</details>

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

chrisbobbe commented:
Thanks, LGTM! Probably Greg should also review the small edits.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2748,7 +2748,40 @@ paths:
+                                This can reflect a direct user action, or can be the indirect
+                                consequence of another action. Whatever the cause, if there's a change
+                                in the set of message flags that the user has for a message, then an
+                                `update_message_flags op: add` event will be sent with the change. Note
```
Presumably `remove` is also possible!

gnprice replied:
a followup: https://github.com/zulip/zulip/pull/27348#issuecomment-1777935699

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2748,7 +2748,40 @@ paths:
+                                  message's sender.
+                                - The `"read"` flag is added after the user unsubscribes from a stream,
+                                  or messages are moved to a not-subscribed stream, provided the user
+                                  can still access the messages at all.
```
We can extend this last bullet to mention that users losing access will get a `delete_message` event instead, just for clarity.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2748,7 +2748,40 @@ paths:
+                                simultaneously. For example, when messages that were unread are moved to
+                                a stream where the user is not subscribed, the resulting change in
+                                message flags (and the corresponding `update_message_flags op: add`
+                                event with flag `"read"`) may happen later than the message move itself.
```
Maybe clearest would be something like this:

"While in most cases, events for changes in message flags resulting from other changes will be sent at the same time as the changes themselves (within hundreds of milliseconds), in some code paths where expensive database operations are involved, a deferred work queue processor is used, which can in rare cases backlog, resulting in a larger delay. For example, ... ('when messages are moved ... with flag `"read"`' goes here) ... uses such a queue processor and thus clients may receive events somewhat than they receive the message move event itself."

gnprice replied:
Hmm, "within hundreds of milliseconds" is not what I'd call simultaneous — from a client's perspective, that is a short while later. In particular it means that the client will be showing the user a UI, for long enough for the user to see it, based on having received the one change but not yet the other.

What are other examples where you'd expect that kind of delay?

gnprice commented:
Thanks!

> Presumably `remove` is also possible! (https://github.com/zulip/zulip/pull/27348#discussion_r1370600154)

Yeah, in my draft in chat I wrote "an `update_message_flags` event". And I see I didn't actually articulate this, but my thinking was:
* we want to describe what causes the `remove` events as well as `add`;
* most of what we'd want to say is exactly the same for both, just flipping around the direction of some of the words;
* so rather than duplicate it (and end up having them get out of sync etc.), let's write one description that covers both. Then the `remove` description can point at the `add` description for the detailed discussion.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -2748,7 +2748,40 @@ paths:
+                                disappears, a [`message`][message-event] or
+                                [`delete_message`][message-delete] event is sent instead.
+
+                                Some examples of actions that trigger an `update_message_flags op: add`
```
I'd prefer to reserve code font for things that appear verbatim somewhere in the API, which isn't true of the string "update_message_flags op: add".

Could be written "`update_message_flags` op: `add`". Or some other wording. I don't think we have a convention we've settled on for expressing a type+op pair in the API docs.

(In chat threads I've been happy writing "an `update_message_flags`/`add` event", with a slash — and note the slash isn't in code font, because only the two pieces appear verbatim in the API, not the whole. In docs that might be a bit cryptic… though in this context it should be self-explanatory enough, and in general I guess whatever spelling we use we can linkify to make it clear.)

laurynmm commented:
**Updates**:
- Revised text for op: `add` event to not mention the op value and linked to that description in the op: `remove` event. Thanks @gnprice for clarifying your thoughts there about this text being used in one place for both op types.
- If the event and op are both mentioned in the text, then the formatting I went with was "`event_name` op: `type`".
- Updated text about delay in events for what was discussed and [drafted on CZO](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/unreads.3A.20messages.20from.20muted.20users.3F/near/1669354).
- Updated the `delete_message` event to note the moved message case, as discussed [here on CZO](https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.3A.20access.20to.20messages/near/1668094).

**Updated screenshots**:

<details>
<summary>Updated update_message_flags op: add</summary>

![Screenshot from 2023-10-26 11-47-26](https://github.com/zulip/zulip/assets/63245456/5464dfd4-d793-4371-8e01-a9998b6ea0db)

</details>
<details>
<summary>Updated update_message_flags op: remove</summary>

![Screenshot from 2023-10-26 11-47-48](https://github.com/zulip/zulip/assets/63245456/7e314400-54ba-42fd-b26b-19dd3b919431)

</details>
<details>
<summary>Updated delete_message event</summary>

![Screenshot from 2023-10-26 11-48-37](https://github.com/zulip/zulip/assets/63245456/580dff83-4103-41c2-9b3f-9fb70eb08ddf)

</details>

laurynmm commented:
**Updates**:
- Revised `delete_message` event for feedback in [#api design conversation on CZO](https://chat.zulip.org/#narrow/stream/378-api-design/topic/unreads.3A.20access.20to.20messages/near/1670647)

chrisbobbe commented:
Thanks! LGTM; over to @gnprice for his review. 🙂

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -2755,7 +2755,44 @@ paths:
+                                message flags (and the corresponding `update_message_flags` event with
+                                flag `"read"`) may happen later than the message move itself. The delay
+                                in that example is typically hundreds of milliseconds and can in rare
+                                cases be minutes or longer.
```
I think the delay there is "typically at most a few hundred milliseconds" -- for streams that are not huge and an empty queue, which is in some sense the most typical state, it'll be <5ms.

timabbott commented:
Posted one tiny proposed tweak but otherwise lgtm.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -2748,7 +2755,44 @@ paths:
+                                - The `"read"` flag is added after the user unsubscribes from a stream,
+                                  or messages are moved to a not-subscribed stream, provided the user
+                                  can still access the messages at all. Note a
+                                  [`delete_message`][message-delete] event is sent in the case the user
+                                  can no longer access the messages.
```
nit:
```suggestion
                                  [`delete_message`][message-delete] event is sent in the case where the user
                                  can no longer access the messages.
```

For me the version without "where" is ungrammatical.

gnprice reviewed (commented):
Thanks! All looks good; one nit below.

laurynmm commented:
**Updates**:
- Revised for @gnprice's [note for a grammar fix](https://github.com/zulip/zulip/pull/27348#discussion_r1376747695) above.
- Marking for integration review.

<details>
<summary>Screenshot of current updated description of update_message_flags op: add event</summary>

![Screenshot from 2023-10-31 11-33-13](https://github.com/zulip/zulip/assets/63245456/fad3f9d3-1573-4a21-b613-f4f15fe1a8a1)

</details>

timabbott commented:
Merged, thanks @laurynmm, @chrisbobbe, and @gnprice!
